File: src/helmlab/utils/gamut.py
"""Gamut mapping — chroma-reduction and adaptive (Ottosson-style) clipping.

Supports sRGB and Display P3 gamuts.

Methods:
    "chroma"   — reduce chroma at fixed L and hue (default, backward compat)
    "adaptive" — Ottosson-style adaptive L0 clipping that shifts both L and C
"""

import numpy as np

from helmlab.utils.srgb_convert import (
    M_XYZ_TO_SRGB,
    M_XYZ_TO_DISPLAYP3,
)

# Gamut → linear-RGB matrix mapping
_GAMUT_MATRICES = {
    "srgb": M_XYZ_TO_SRGB,
    "display-p3": M_XYZ_TO_DISPLAYP3,
}


def _linear_rgb(XYZ: np.ndarray, gamut: str) -> np.ndarray:
    """XYZ → linear RGB for the given gamut (no gamma)."""
    M = _GAMUT_MATRICES[gamut]
    return np.asarray(XYZ, dtype=np.float64) @ M.T


def is_in_gamut(lab: np.ndarray, space, gamut: str = "srgb", tol: float = 1e-4) -> np.ndarray:
    """Check whether Lab coordinates are inside the specified gamut.

    Parameters
    ----------
    lab : ndarray, shape (..., 3)
    space : ColorSpace instance (must have .to_XYZ)
    gamut : "srgb" or "display-p3"
    tol : tolerance for boundary inclusion

    Returns
    -------
    bool ndarray, shape (...)
    """
    lab = np.asarray(lab, dtype=np.float64)
    XYZ = space.to_XYZ(lab)
    rgb = _linear_rgb(XYZ, gamut)
    return np.all((rgb >= -tol) & (rgb <= 1.0 + tol), axis=-1)


def max_chroma(L: float, H_rad: float, space, gamut: str = "srgb", tol: float = 1e-4) -> float:
    """Binary search for maximum in-gamut chroma at fixed L and H.

    Parameters
    ----------
    L : lightness value
    H_rad : hue angle in radians
    space : ColorSpace with to_XYZ method
    gamut : "srgb" or "display-p3"
    tol : convergence tolerance

    Returns
    -------
    float — maximum chroma that stays in gamut
    """
    cos_h = np.cos(H_rad)
    sin_h = np.sin(H_rad)

    lo, hi = 0.0, 1.0

    # First expand hi until it's out of gamut
    lab_test = np.array([L, hi * cos_h, hi * sin_h])
    while is_in_gamut(lab_test, space, gamut, tol):
        hi *= 2.0
        if hi > 100.0:  # safety
            return hi
        lab_test = np.array([L, hi * cos_h, hi * sin_h])

    # Binary search
    for _ in range(50):
        mid = (lo + hi) * 0.5
        lab_test = np.array([L, mid * cos_h, mid * sin_h])
        if is_in_gamut(lab_test, space, gamut, tol):
            lo = mid
        else:
            hi = mid
        if hi - lo < tol:
            break

    return lo


def find_cusp(H_rad: float, space, gamut: str = "srgb",
              n_scan: int = 64, tol: float = 1e-4) -> tuple[float, float]:
    """Find the cusp (L_cusp, C_cusp) at a given hue angle.

    The cusp is the point on the gamut boundary where chroma is maximized.
    Uses a coarse scan followed by golden-section refinement.

    Parameters
    ----------
    H_rad : hue angle in radians
    space : ColorSpace with to_XYZ method
    gamut : "srgb" or "display-p3"
    n_scan : number of L values in coarse scan
    tol : convergence tolerance for L

    Returns
    -------
    (L_cusp, C_cusp) — lightness and chroma at the cusp
    """
    # Coarse scan: find L that gives maximum chroma
    # Skip degenerate values where max_chroma returns unbounded (>10)
    L_vals = np.linspace(0.05, 0.98, n_scan)
    best_L = 0.5
    best_C = 0.0
    for L in L_vals:
        C = max_chroma(L, H_rad, space, gamut, tol)
        if C > 10.0:
            # Degenerate: gamut appears unbounded at this L, skip
            continue
        if C > best_C:
            best_C = C
            best_L = L

    # Golden-section refinement around the best L
    golden = (np.sqrt(5) - 1) / 2
    idx = int(np.argmin(np.abs(L_vals - best_L)))
    a = L_vals[max(0, idx - 2)]
    b = L_vals[min(len(L_vals) - 1, idx + 2)]

    for _ in range(30):
        if b - a < tol:
            break
        c = b - golden * (b - a)
        d = a + golden * (b - a)
        Cc = max_chroma(c, H_rad, space, gamut, tol)
        Cd = max_chroma(d, H_rad, space, gamut, tol)
        # Treat degenerate values as zero for comparison
        if Cc > 10.0:
            Cc = 0.0
        if Cd > 10.0:
            Cd = 0.0
        if Cc > Cd:
            b = d
        else:
            a = c

    L_cusp = (a + b) / 2
    C_cusp = max_chroma(L_cusp, H_rad, space, gamut, tol)
    if C_cusp > 10.0:
        C_cusp = best_C  # fallback to coarse scan result
        L_cusp = best_L
    return L_cusp, C_cusp


def _gamut_clip_adaptive_single(lab: np.ndarray, space, gamut: str,
                                 alpha: float) -> np.ndarray:
    """Adaptive gamut clipping for a single Lab triplet (Ottosson-style).

    Instead of only reducing chroma (which preserves L exactly),
    this method adaptively shifts both L and C toward the gamut boundary.

    The projection target L0 is computed adaptively:
    - For colors near mid-lightness: mostly chroma reduction
    - For very light/dark out-of-gamut colors: allows L to shift
    - alpha controls the blend: 0 = pure chroma reduction, larger = more L shift

    This is especially important for yellow where the cusp is at very high L.
    Pure chroma reduction at high L gives near-zero chroma (the cliff problem).
    Adaptive clipping can shift L down toward the cusp, preserving chroma.
    """
    if is_in_gamut(lab, space, gamut):
        return lab.copy()

    L = float(lab[0])
    a, b = float(lab[1]), float(lab[2])
    C = np.sqrt(a ** 2 + b ** 2)
    H = np.arctan2(b, a)

    if C < 1e-10:
        # Achromatic but OOG — clamp L to achievable range
        result = lab.copy()
        result[1] = 0.0
        result[2] = 0.0
        return result

    # Find cusp at this hue
    L_cusp, C_cusp = find_cusp(H, space, gamut)

    # Compute adaptive L0 (Ottosson's formula adapted for [0,1] L range)
    Ld = L - L_cusp
    if Ld >= 0:
        k = 2.0 * (1.0 - L_cusp)
    else:
        k = 2.0 * L_cusp

    # Guard against degenerate k (cusp at extreme L)
    k = max(k, 1e-6)

    e1 = k / 2.0 + abs(Ld) + alpha * C / k
    discriminant = e1 * e1 - 2.0 * k * abs(Ld)

    if discriminant < 0:
        # Fallback: L0 = L_cusp
        L0 = L_cusp
    else:
        sgn = 1.0 if Ld >= 0 else -1.0
        L0 = L_cusp + sgn * (e1 - np.sqrt(discriminant)) / 2.0

    # Clamp L0 to valid range
    L0 = float(np.clip(L0, 0.0, 1.0))

    # Binary search for t in [0, 1] on the line from (L0, 0) to (L, C)
    # such that the point (L0*(1-t) + t*L, t*C) is on the gamut boundary
    cos_h = np.cos(H)
    sin_h = np.sin(H)

    lo_t, hi_t = 0.0, 1.0

    for _ in range(50):
        t = (lo_t + hi_t) * 0.5
        L_test = L0 * (1.0 - t) + t * L
        C_test = t * C
        lab_test = np.array([L_test, C_test * cos_h, C_test * sin_h])
        if is_in_gamut(lab_test, space, gamut):
            lo_t = t
        else:
            hi_t = t
        if hi_t - lo_t < 1e-6:
            break

    # Use the last in-gamut value
    t = lo_t
    L_clip = L0 * (1.0 - t) + t * L
    C_clip = t * C
    return np.array([L_clip, C_clip * cos_h, C_clip * sin_h])


def gamut_map(lab: np.ndarray, space, gamut: str = "srgb", method: str = "chroma",
              alpha: float = 0.05) -> np.ndarray:
    """Map Lab coordinates into the specified gamut.

    Parameters
    ----------
    lab : ndarray, shape (3,) or (N, 3)
    space : ColorSpace with to_XYZ method
    gamut : "srgb" or "display-p3"
    method : "chroma" or "adaptive"
        "chroma" — reduce chroma while preserving L and hue (default)
        "adaptive" — Ottosson-style adaptive clipping that shifts both L and C
    alpha : float
        Blend parameter for adaptive method (default 0.05).
        0 = pure chroma reduction, larger = more L shift allowed.
        Ignored when method="chroma".

    Returns
    -------
    ndarray, same shape as input — gamut-mapped Lab
    """
    lab = np.asarray(lab, dtype=np.float64)
    if lab.ndim == 1:
        return _gamut_map_single(lab, space, gamut, method=method, alpha=alpha)
    return gamut_map_batch(lab, space, gamut, method=method, alpha=alpha)


def _gamut_map_single(lab: np.ndarray, space, gamut: str,
                      method: str = "chroma", alpha: float = 0.05) -> np.ndarray:
    """Gamut-map a single Lab triplet.

    Dispatches to chroma reduction or adaptive clipping based on method.
    """
    if method == "adaptive":
        return _gamut_clip_adaptive_single(lab, space, gamut, alpha)

    # Default: chroma reduction
    if is_in_gamut(lab, space, gamut):
        return lab.copy()

    L = lab[0]
    a, b = lab[1], lab[2]
    C = np.sqrt(a ** 2 + b ** 2)
    H = np.arctan2(b, a)

    if C < 1e-10:
        # Achromatic but OOG — clamp L to achievable range
        result = lab.copy()
        result[1] = 0.0
        result[2] = 0.0
        return result

    C_max = max_chroma(L, H, space, gamut)
    C_new = min(C, C_max)
    return np.array([L, C_new * np.cos(H), C_new * np.sin(H)])


def gamut_map_batch(labs: np.ndarray, space, gamut: str = "srgb",
                    method: str = "chroma", alpha: float = 0.05) -> np.ndarray:
    """Batch gamut mapping.

    Parameters
    ----------
    labs : ndarray, shape (N, 3)
    space : ColorSpace with to_XYZ method
    gamut : "srgb" or "display-p3"
    method : "chroma" or "adaptive"
    alpha : blend parameter for adaptive method

    Returns
    -------
    ndarray, shape (N, 3) — gamut-mapped Lab values
    """
    labs = np.asarray(labs, dtype=np.float64)
    in_gamut_mask = is_in_gamut(labs, space, gamut)
    result = labs.copy()

    oog_idx = np.where(~in_gamut_mask)[0]
    for i in oog_idx:
        result[i] = _gamut_map_single(labs[i], space, gamut,
                                       method=method, alpha=alpha)

    return result

File: packages/helmlab-js/src/utils/gamut.ts
/** Gamut mapping — binary search chroma reduction preserving L and hue. */

import type { Lab, XYZ } from '../types.js';
import { M_XYZ_TO_SRGB, M_XYZ_TO_DISPLAYP3, M_XYZ_TO_REC2020 } from './srgb.js';

const { cos, sin, sqrt, atan2, min } = Math;

/** Minimal interface for gamut mapping — any object with toXYZ. */
export interface SpaceLike {
  toXYZ(lab: Lab): XYZ;
}

export type Gamut = 'srgb' | 'display-p3' | 'rec2020';

function getMatrix(gamut: Gamut): Float64Array {
  if (gamut === 'srgb') return M_XYZ_TO_SRGB;
  if (gamut === 'display-p3') return M_XYZ_TO_DISPLAYP3;
  return M_XYZ_TO_REC2020;
}

/** Check if XYZ is in the given RGB gamut (linear check, no gamma). */
function xyzInGamut(M: Float64Array, x: number, y: number, z: number, tol: number): boolean {
  const r = M[0] * x + M[1] * y + M[2] * z;
  const g = M[3] * x + M[4] * y + M[5] * z;
  const b = M[6] * x + M[7] * y + M[8] * z;
  return r >= -tol && r <= 1 + tol &&
         g >= -tol && g <= 1 + tol &&
         b >= -tol && b <= 1 + tol;
}

/** Check if Lab coordinates are in the specified gamut. */
export function isInGamut(lab: Lab, space: SpaceLike, gamut: Gamut = 'srgb', tol = 1e-4): boolean {
  const xyz = space.toXYZ(lab);
  return xyzInGamut(getMatrix(gamut), xyz[0], xyz[1], xyz[2], tol);
}

/** Binary search for maximum in-gamut chroma at fixed L and hue. */
export function maxChroma(L: number, hRad: number, space: SpaceLike, gamut: Gamut = 'srgb', tol = 1e-4): number {
  const cosH = cos(hRad);
  const sinH = sin(hRad);
  const M = getMatrix(gamut);

  let lo = 0, hi = 1;

  // Expand hi until out of gamut
  while (true) {
    const xyz = space.toXYZ([L, hi * cosH, hi * sinH]);
    if (!xyzInGamut(M, xyz[0], xyz[1], xyz[2], tol)) break;
    hi *= 2;
    if (hi > 100) return hi;
  }

  // Binary search
  for (let i = 0; i < 50; i++) {
    const mid = (lo + hi) * 0.5;
    const xyz = space.toXYZ([L, mid * cosH, mid * sinH]);
    if (xyzInGamut(M, xyz[0], xyz[1], xyz[2], tol)) lo = mid; else hi = mid;
    if (hi - lo < tol) break;
  }

  return lo;
}

/** Find the cusp (L, C) at a given hue angle — maximum chroma over all L values. */
export function findCusp(hRad: number, space: SpaceLike, gamut: Gamut = 'srgb', nScan = 64, tol = 1e-4): [number, number] {
  const cosH = cos(hRad);
  const sinH = sin(hRad);
  let bestL = 0.5, bestC = 0;

  // Coarse scan
  for (let i = 0; i <= nScan; i++) {
    const L = i / nScan;
    const C = maxChroma(L, hRad, space, gamut, tol);
    if (C > bestC) { bestC = C; bestL = L; }
  }

  // Fine refine around bestL
  const step = 1 / nScan;
  let lo = Math.max(0, bestL - step), hi = Math.min(1, bestL + step);
  for (let i = 0; i < 20; i++) {
    const m1 = lo + (hi - lo) / 3;
    const m2 = hi - (hi - lo) / 3;
    const c1 = maxChroma(m1, hRad, space, gamut, tol);
    const c2 = maxChroma(m2, hRad, space, gamut, tol);
    if (c1 < c2) lo = m1; else hi = m2;
  }
  bestL = (lo + hi) / 2;
  bestC = maxChroma(bestL, hRad, space, gamut, tol);

  return [bestL, bestC];
}

/** Gamut-map a single Lab by reducing chroma (preserving L and hue). */
function gamutMapSingle(lab: Lab, space: SpaceLike, gamut: Gamut): Lab {
  if (isInGamut(lab, space, gamut)) return [...lab];

  const [L, a, b] = lab;
  const C = sqrt(a * a + b * b);
  const H = atan2(b, a);

  if (C < 1e-10) return [L, 0, 0];

  const Cmax = maxChroma(L, H, space, gamut);
  const Cn = min(C, Cmax);
  return [L, Cn * cos(H), Cn * sin(H)];
}

/** Adaptive gamut clipping (Ottosson-style) — mirrors the Python sibling's
 * `_gamut_clip_adaptive_single`. Instead of only reducing chroma, shifts
 * both L and C toward a cusp-derived projection target: important for
 * hues like yellow whose cusp sits at very high L, where pure chroma
 * reduction collapses to near-gray (the cliff problem). */
function gamutClipAdaptiveSingle(lab: Lab, space: SpaceLike, gamut: Gamut, alpha: number): Lab {
  if (isInGamut(lab, space, gamut)) return [...lab];

  const [L, a, b] = lab;
  const C = sqrt(a * a + b * b);
  const H = atan2(b, a);

  if (C < 1e-10) return [L, 0, 0];

  const [Lcusp] = findCusp(H, space, gamut);

  const Ld = L - Lcusp;
  let k = Ld >= 0 ? 2 * (1 - Lcusp) : 2 * Lcusp;
  k = Math.max(k, 1e-6);

  const e1 = k / 2 + Math.abs(Ld) + (alpha * C) / k;
  const discriminant = e1 * e1 - 2 * k * Math.abs(Ld);

  let L0: number;
  if (discriminant < 0) {
    L0 = Lcusp;
  } else {
    const sgn = Ld >= 0 ? 1 : -1;
    L0 = Lcusp + (sgn * (e1 - sqrt(discriminant))) / 2;
  }
  L0 = min(Math.max(L0, 0), 1);

  const cosH = cos(H);
  const sinH = sin(H);

  let loT = 0;
  let hiT = 1;
  for (let i = 0; i < 50; i++) {
    const t = (loT + hiT) * 0.5;
    const Ltest = L0 * (1 - t) + t * L;
    const Ctest = t * C;
    if (isInGamut([Ltest, Ctest * cosH, Ctest * sinH], space, gamut)) {
      loT = t;
    } else {
      hiT = t;
    }
    if (hiT - loT < 1e-6) break;
  }

  const t = loT;
  const Lclip = L0 * (1 - t) + t * L;
  const Cclip = t * C;
  return [Lclip, Cclip * cosH, Cclip * sinH];
}

export type GamutMapMethod = 'chroma' | 'adaptive';

/** Gamut-map Lab coordinates.
 * `method='chroma'` (default) reduces chroma at constant L and hue;
 * `method='adaptive'` is the Ottosson-style cusp projection. */
export function gamutMap(lab: Lab, space: SpaceLike, gamut: Gamut = 'srgb',
                         method: GamutMapMethod = 'chroma', alpha = 0.05): Lab {
  if (method === 'adaptive') return gamutClipAdaptiveSingle(lab, space, gamut, alpha);
  return gamutMapSingle(lab, space, gamut);
}
